<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ Kandinsky</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Kandinsky</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/modC.html">Kandinsky cell/spot-neighbour based comparison</a>
    </li>
    <li>
      <a href="../articles/modD.html">Kandinsky cell/spot co-localisation/dispersion</a>
    </li>
    <li>
      <a href="../articles/modE.html">Kandinsky hot/cold areas</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ciccalab/Kandinsky/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="AddTxPath.html">AddTxPath()</a></code> </p>
        </td>
        <td><p>Add transcript file path to Kandinsky data in a Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="AlignXeniumCoords.html">AlignXeniumCoords()</a></code> </p>
        </td>
        <td><p>align Xenium image and cell coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="CosmxFovPlot.html">CosmxFovPlot()</a></code> </p>
        </td>
        <td><p>Visualize CosMx IFs staining for a list of FOVs</p></td>
      </tr><tr><td>
          <p><code><a href="EnrichPlots.html">EnrichPlots()</a></code> </p>
        </td>
        <td><p>Combined PropPlot and ResidualPlot</p></td>
      </tr><tr><td>
          <p><code><a href="GetTxData.html">GetTxData()</a></code> </p>
        </td>
        <td><p>Query and filter Kandinsky transcript file</p></td>
      </tr><tr><td>
          <p><code><a href="KanData-set.html">`KanData&lt;-`()</a></code> </p>
        </td>
        <td><p>Assign data to Kandinsky</p></td>
      </tr><tr><td>
          <p><code><a href="KanData.html">KanData()</a></code> <code><a href="KanData.html">`KanData&lt;-`(<i>&lt;Kandinsky&gt;</i>)</a></code> <code><a href="KanData.html">`KanData&lt;-`(<i>&lt;Seurat&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="KanPlot.html">KanPlot()</a></code> </p>
        </td>
        <td><p>visualize Kandinsky and Seurat spatial data</p></td>
      </tr><tr><td>
          <p><code><a href="Kandinsky-class.html">Kandinsky-class</a></code> <code><a href="Kandinsky-class.html">Kandinsky</a></code> </p>
        </td>
        <td><p>Kandinsky S4 object class</p></td>
      </tr><tr><td>
          <p><code><a href="LoadXeniumImage.html">LoadXeniumImage()</a></code> </p>
        </td>
        <td><p>Load Xenium OME tiff image</p></td>
      </tr><tr><td>
          <p><code><a href="PackedSpatRaster.html">PackedSpatRaster()</a></code> </p>
        </td>
        <td><p>S4 Class Union KanImg</p></td>
      </tr><tr><td>
          <p><code><a href="PropPlot.html">PropPlot()</a></code> </p>
        </td>
        <td><p>Proportion barplot</p></td>
      </tr><tr><td>
          <p><code><a href="ResidualPlot.html">ResidualPlot()</a></code> </p>
        </td>
        <td><p>Chi-square test residual plot</p></td>
      </tr><tr><td>
          <p><code><a href="SlidePlot.html">SlidePlot()</a></code> </p>
        </td>
        <td><p>Plot Kandinsky images</p></td>
      </tr><tr><td>
          <p><code><a href="addKanData.html">addKanData()</a></code> </p>
        </td>
        <td><p>Add Kandinsky data from .rds file to Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="c12.html">c12</a></code> </p>
        </td>
        <td><p>C12 discrete color palette</p></td>
      </tr><tr><td>
          <p><code><a href="c24.html">c24</a></code> </p>
        </td>
        <td><p>C24 discrete color palette</p></td>
      </tr><tr><td>
          <p><code><a href="centroid_nb.html">centroid_nb()</a></code> </p>
        </td>
        <td><p>centroid distance neighbour identification</p></td>
      </tr><tr><td>
          <p><code><a href="fireflyPlot.html">fireflyPlot()</a></code> </p>
        </td>
        <td><p>Plot minimum distance between cell types as a spatial heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="gene_contamination_score.html">gene_contamination_score()</a></code> </p>
        </td>
        <td><p>Calculate gene-specific contamination score</p></td>
      </tr><tr><td>
          <p><code><a href="get_nbcounts.html">get_nbcounts()</a></code> </p>
        </td>
        <td><p>Get neighbourhood aggregated gene expression count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="get_visium_textures.html">get_visium_textures()</a></code> </p>
        </td>
        <td><p>Extract texture features from Visium image</p></td>
      </tr><tr><td>
          <p><code><a href="global_biv_spatcor.html">global_biv_spatcor()</a></code> </p>
        </td>
        <td><p>Compute global bivariate spatial correlation</p></td>
      </tr><tr><td>
          <p><code><a href="global_univ_spatcor.html">global_univ_spatcor()</a></code> </p>
        </td>
        <td><p>Compute global Moran's I spatial autocorrelation statistic</p></td>
      </tr><tr><td>
          <p><code><a href="he_mask.html">he_mask()</a></code> </p>
        </td>
        <td><p>Mask Visium H&amp;E image</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_analysis.html">hotspot_analysis()</a></code> </p>
        </td>
        <td><p>Hotspot analysis using Getis-Ord Gi* statistics</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_count.html">hotspot_count()</a></code> </p>
        </td>
        <td><p>Count and describe Getis-Ord Gi* hotspot sites</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_overlap.html">hotspot_overlap()</a></code> </p>
        </td>
        <td><p>define overlap between hot/cold spots of two variables</p></td>
      </tr><tr><td>
          <p><code><a href="jc_coloc.html">jc_coloc()</a></code> </p>
        </td>
        <td><p>JC colocalization test and plot using Seurat/Kandinsky object</p></td>
      </tr><tr><td>
          <p><code><a href="kandinsky_init.html">kandinsky_init()</a></code> </p>
        </td>
        <td><p>Initialize Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="knn_nb.html">knn_nb()</a></code> </p>
        </td>
        <td><p>K-nearest neighbour neighbour identification</p></td>
      </tr><tr><td>
          <p><code><a href="load_g4x_img.html">load_g4x_img()</a></code> </p>
        </td>
        <td><p>Create spatraster image from G4X jp2 image file</p></td>
      </tr><tr><td>
          <p><code><a href="membrane_nb.html">membrane_nb()</a></code> </p>
        </td>
        <td><p>membrane-based neighbour identification</p></td>
      </tr><tr><td>
          <p><code><a href="multi_jc.html">multi_jc()</a></code> </p>
        </td>
        <td><p>same-color (BB) and different-color (BW) join count test</p></td>
      </tr><tr><td>
          <p><code><a href="nbCluster.html">nbCluster()</a></code> </p>
        </td>
        <td><p>Identify single cell/spot clusters on the basis of their neighbourhood compositin</p></td>
      </tr><tr><td>
          <p><code><a href="nbSizePlot.html">nbSizePlot()</a></code> </p>
        </td>
        <td><p>density plot of neighbourhood size across cells/spots</p></td>
      </tr><tr><td>
          <p><code><a href="nb_entropy.html">nb_entropy()</a></code> </p>
        </td>
        <td><p>Calculate Shannon entropy of cell neighbourhood composition</p></td>
      </tr><tr><td>
          <p><code><a href="nb_expand.html">nb_expand()</a></code> </p>
        </td>
        <td><p>expand neighbour links to N higher orders</p></td>
      </tr><tr><td>
          <p><code><a href="nb_lm.html">nb_lm()</a></code> </p>
        </td>
        <td><p>Neighbourhood-based multiple linear regression of gene expression values</p></td>
      </tr><tr><td>
          <p><code><a href="nb_summary.html">nb_summary()</a></code> </p>
        </td>
        <td><p>Get Kandinsky neighbourhood summary metrics</p></td>
      </tr><tr><td>
          <p><code><a href="nb_update.html">nb_update()</a></code> </p>
        </td>
        <td><p>Update definition of Kandinsky neighbour network</p></td>
      </tr><tr><td>
          <p><code><a href="nnMat.html">nnMat()</a></code> </p>
        </td>
        <td><p>Neighbour composition matrix</p></td>
      </tr><tr><td>
          <p><code><a href="nn_query.html">nn_query()</a></code> </p>
        </td>
        <td><p>Interrogate neighbour matrix</p></td>
      </tr><tr><td>
          <p><code><a href="populate_sf.html">populate_sf()</a></code> </p>
        </td>
        <td><p>Add variables to sf object stored in Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_cosmx_seurat.html">prepare_cosmx_seurat()</a></code> </p>
        </td>
        <td><p>Prepare NanoString Cosmx Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_g4x_seurat.html">prepare_g4x_seurat()</a></code> </p>
        </td>
        <td><p>Prepare Singular Genomics G4X Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_merscope_seurat.html">prepare_merscope_seurat()</a></code> </p>
        </td>
        <td><p>Prepare Vizgen Merscope Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_proseg_seurat.html">prepare_proseg_seurat()</a></code> </p>
        </td>
        <td><p>Prepare proseg Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_seurat_other.html">prepare_seurat_other()</a></code> </p>
        </td>
        <td><p>prepare seurat object starting from any spatial transcriptomic/proteomic data in tabular format</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_slideseq_seurat.html">prepare_slideseq_seurat()</a></code> </p>
        </td>
        <td><p>Prepare Slide-Seq Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_visiumHD_seurat.html">prepare_visiumHD_seurat()</a></code> </p>
        </td>
        <td><p>Prepare 10X Visium HD Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_visium_seurat.html">prepare_visium_seurat()</a></code> </p>
        </td>
        <td><p>Prepare 10X Visium Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_xenium_seurat.html">prepare_xenium_seurat()</a></code> </p>
        </td>
        <td><p>Prepare 10X Xenium Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="queen_nb.html">queen_nb()</a></code> </p>
        </td>
        <td><p>queen contiguity neighbour identification</p></td>
      </tr><tr><td>
          <p><code><a href="readKanData.html">readKanData()</a></code> </p>
        </td>
        <td><p>Read Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="read_fovfile.html">read_fovfile()</a></code> </p>
        </td>
        <td><p>Read CosMx fov position file</p></td>
      </tr><tr><td>
          <p><code><a href="resample_cells.html">resample_cells()</a></code> </p>
        </td>
        <td><p>resample single cells in Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="saveKanData.html">saveKanData()</a></code> </p>
        </td>
        <td><p>Save Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="scale_coords.html">scale_coords()</a></code> </p>
        </td>
        <td><p>Rescale Kandinsky polygon xy coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="smi2sf.html">smi2sf()</a></code> </p>
        </td>
        <td><p>Create sf geometry object from CosMx Seurat data</p></td>
      </tr><tr><td>
          <p><code><a href="spweights_mat.html">spweights_mat()</a></code> </p>
        </td>
        <td><p>estimate weight summary constansts from spatial weights matrix</p></td>
      </tr><tr><td>
          <p><code><a href="stitch_samples.html">stitch_samples()</a></code> </p>
        </td>
        <td><p>Merge coordinate reference systems from independent samples</p></td>
      </tr><tr><td>
          <p><code><a href="subsample_nb.html">subsample_nb()</a></code> </p>
        </td>
        <td><p>subsample neighbour network links</p></td>
      </tr><tr><td>
          <p><code><a href="tri_nb.html">tri_nb()</a></code> </p>
        </td>
        <td><p>Delaunay triangulation neighbour identification</p></td>
      </tr><tr><td>
          <p><code><a href="update_kandinsky.html">update_kandinsky()</a></code> </p>
        </td>
        <td><p>Update Kandinsky data</p></td>
      </tr><tr><td>
          <p><code><a href="visium2rast.html">visium2rast()</a></code> </p>
        </td>
        <td><p>Create spatraster image from Visium image file</p></td>
      </tr><tr><td>
          <p><code><a href="visium2sf.html">visium2sf()</a></code> </p>
        </td>
        <td><p>Create sf geometry object from Visium Seurat data</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pietro Andrei, Mariachiara Grieco, Matteo Cereda, Francesca Ciccarelli.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

